File: chapter1/002/main.go
package main

import "fmt"

func main() {
	fmt.Println(isPermutation("", ""))
	fmt.Println(isPermutation("abba", "aabb"))
	fmt.Println(isPermutation("stẽnio", "ẽinost"))
	fmt.Println(isPermutation("arara", "aarra"))

	fmt.Println(isPermutation("", "aarra"))
	fmt.Println(isPermutation("arara", "arera"))
}

func isPermutation(s1, s2 string) bool {
	runes := make(map[rune]int)

	for _, r := range s1 {
		runes[r]++
	}

	for _, r := range s2 {
		runes[r]--
	}

	for _, v := range runes {
		if v != 0 {
			return false
		}
	}

	return true
}

File: chapter1/002/main.py
def isPermutation(s1, s2):
	chars = {}

	for c in s1:
	    chars[c] = chars.get(c, 0) + 1

	for c in s2:
	    chars[c] = chars.get(c, 0) - 1

	for _, v in chars.items():
		if v != 0:
			return False

	return True

print(isPermutation("", ""))
print(isPermutation("abba", "aabb"))
print(isPermutation("stẽnio", "ẽinost"))
print(isPermutation("arara", "aarra"))

print(isPermutation("", "aarra"))
print(isPermutation("arara", "arera"))